dpd/src/mcast.rs

```diff
@@ -0,0 +1,1271 @@
+                    s,
+                    ipv6,
+                    group_info.nat_target.unwrap(),
+                );
+            }
+
+            if res.is_ok() {
+                res = table::mcast::route::add_ipv6_entry(
+                    s,
+                    ipv6,
+                    group_info.vlan_id,
+                );
+            }
+
+            res
+        }
+    };
+
+    if let Err(e) = result {
+        cleanup_on_group_create(
+            s,
+            group_ip,
+            group_id,
+            &added_members,
+            group_info.nat_target,
+            group_info.sources.as_deref(),
+        )?;
+        return Err(e);
+    }
+
+    // Only store configuration if all operations succeeded
+    let group = MulticastGroup {
+        group_id,
+        tag: group_info.tag,
+        int_fwding: InternalForwarding {
+            nat_target: group_info.nat_target,
+        },
+        ext_fwding: ExternalForwarding {
+            vlan_id: group_info.vlan_id,
+        },
+        sources: group_info.sources,
+        replication_info: MulticastReplicationInfo {
+            replication_id: rid,
+            level1_excl_id,
+            level2_excl_id,
+        },
+        members: group_info.members,
+    };
+
+    mcast.groups.insert(group_ip, group.clone());
+    mcast.used_group_ids.insert(group_id);
+
+    Ok(MulticastGroupResponse::new(group_ip, &group))
+}
+
+/// Delete a multicast group from the switch, including all associated tables
+/// and port mappings.
+pub(crate) fn del_group(s: &Switch, ident: Identifier) -> DpdResult<()> {
+    let (group_ip, group) = find_group(s, &ident)?;
+
+    debug!(s.log, "deleting multicast group for IP {}", group_ip);
+
+    // Now we have both the IP and the group, continue with deletion
+    // Remove the route from all associated tables
+    del_entry(s, group_ip, &group)?;
+
+    // Delete the group from the ASIC, which also deletes the associated ports
+    s.asic_hdl.mc_group_destroy(group.group_id).map_err(|e| {
+        DpdError::McastGroup(format!(
+            "failed to delete multicast group for IP {} with ID {}: {:?}",
+            group_ip, group.group_id, e
+        ))
+    })?;
+
+    // Remove from our tracking
+    let mut mcast = s.mcast.lock().unwrap();
+    mcast.groups.remove(&group_ip);
+
+    Ok(())
+}
+
+/// Get a multicast group configuration.
+pub(crate) fn get_group(
+    s: &Switch,
+    ident: Identifier,
+) -> DpdResult<MulticastGroupResponse> {
+    let (group_ip, group) = find_group(s, &ident)?;
+
+    // Convert to response
+    Ok(MulticastGroupResponse::new(group_ip, &group))
+}
+
+/// Modify a multicast group configuration.
+pub(crate) fn modify_group(
+    s: &Switch,
+    ident: Identifier,
+    new_group_info: MulticastGroupUpdateEntry,
+) -> DpdResult<MulticastGroupResponse> {
+    let (group_ip, group) = find_group(s, &ident)?;
+
+    debug!(s.log, "modifying multicast group for IP {}", group_ip);
```

Comment:
We're touching a few tables in a row here, using deltas computed at the start. Is there a risk that concurrent calls to this endpoint from, e.g., our 3 Nexus instances in the rack might try to insert/remove entries in an inconsistent way?